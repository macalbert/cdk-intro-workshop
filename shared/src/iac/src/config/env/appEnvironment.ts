export enum AppEnvironment {
    Production = "Production",
    Development = "Development",
    Staging = "Staging",
    GithubAction = "GithubAction",
    Test = "Test",
}
