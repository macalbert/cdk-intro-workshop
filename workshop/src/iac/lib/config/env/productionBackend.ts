import {
    ApiGatewayStack,
    type ApiGatewayStackProps,
    AppEnvironment,
    ApiLambdaStack,
    type ApiLambdaStackProps,
    S3Stack,
    type S3StackProps,
    StackBuildPart,
    type StackBuildProps,
} from "@m47/shared-iac";
import { type Stack } from "aws-cdk-lib";
import { SubnetType } from "aws-cdk-lib/aws-ec2";
import type { IFunction } from "aws-cdk-lib/aws-lambda";

export interface ProductionBackendProps extends StackBuildProps {
    clusterName: string;
    serviceName: string;
    exposedPort: number;
    sourceCodePath: string;
    dockerfileApi: string;
}

export class ProductionBackend extends StackBuildPart {
    props: ProductionBackendProps;

    environment = AppEnvironment.Production;

    constructor(buildProps: ProductionBackendProps) {
        super(buildProps);

        this.props = buildProps;
        this.props.githubRepo = this.formatRepoNameForCloudFormation();
    }

    build(): Stack[] {
        
        const apiLambdaStack = this.createApiLambdaStack(
            "ApiLambda",
            this.props.dockerfileApi,
            "api",
        );

        const apiGatewayStack = this.createApiGateway(
            apiLambdaStack.lambdaFunction,
            "ApiGateway",
            "api-cdk-workshop",
        );

        return [
            apiLambdaStack,
            apiGatewayStack,
        ];
    }

    private createApiLambdaStack(
        name: string,
        dockerfile: string,
        subdomain: string,
    ): ApiLambdaStack {
        const lambdaExportProps: ApiLambdaStackProps = {
            name: name,
            githubRepo: this.props.githubRepo,
            pathDockerFile: this.props.sourceCodePath,
            env: this.props.env,
            vpc: this.props.vpc,
            envName: AppEnvironment.Production,
            timeoutSeconds: 29,
            memorySizeMbs: 1024,
            vpcSubnets: {
                subnetType: SubnetType.PRIVATE_WITH_EGRESS,
            },
            stackName: `${this.props.githubRepo}-${name}`,
            functionName: `${this.props.githubRepo}-${subdomain}`.toLowerCase(),
            dockerFile: dockerfile,
        };

        const stack = new ApiLambdaStack(this.props.scope, lambdaExportProps);

        return stack;
    }

    private createApiGateway(
        lambdaFunction: IFunction,
        name: string,
        subdomain: string,
    ): ApiGatewayStack {
        const apiProps: ApiGatewayStackProps = {
            env: this.props.env,
            name: name,
            envName: AppEnvironment.Production,
            lambdaFunction: lambdaFunction,
            stackName: `${this.props.githubRepo}-${name}`,
            subdomain: subdomain,
            githubRepo: this.props.githubRepo,
            certificateArn: `arn:aws:acm:us-east-1:${this.props.env.account}:certificate/be63062d-5316-47af-9f94-819c1dc02853`,
            domain: "m47.io",
        };

        return new ApiGatewayStack(this.props.scope, apiProps);
    }
}
