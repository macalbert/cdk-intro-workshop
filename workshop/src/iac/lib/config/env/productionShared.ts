import {
    AppEnvironment,
    PipelineStack,
    type PipelineStackProps,
    RdsStack,
    type RdsStackProps,
    StackBuildPart,
    type StackBuildProps,
    WindowsBastionStack,
    type WindowsBastionStackProps,
} from "@m47/shared-iac";
import { RemovalPolicy, type Stack } from "aws-cdk-lib";
import { InstanceType, InstanceClass, InstanceSize } from "aws-cdk-lib/aws-ec2";
import { PostgresEngineVersion } from "aws-cdk-lib/aws-rds";

export class ProductionShared extends StackBuildPart {
    props: StackBuildProps;

    constructor(buildProps: StackBuildProps) {
        super(buildProps);

        this.props = buildProps;
    }

    build(): Stack[] {
        const pipelineStack = this.createPipelineStack();
        const rdsStack = this.createRdsStack();
        const createBastionStack = this.createBastionStack();

        return [
            rdsStack,
            pipelineStack,
            createBastionStack,
        ];
    }

    private createPipelineStack(): PipelineStack {
        const pipeline: PipelineStackProps = {
            name: "Pipeline",
            branch: this.props.branch,
            githubRepo: this.props.githubRepo,
            envName: AppEnvironment.Production,
            secretTokenArn: `arn:aws:secretsmanager:${this.props.env.region}:${this.props.env.account}:secret:github-access-token-secret-dth2kY`,
            testBuildSpec: [
                "workshop/src/iac/buildspecs/test/iac.yml",
                "workshop/src/iac/buildspecs/test/backend.yml",
            ],
            deployBuildSpec: [
                "workshop/src/iac/buildspecs/production/backendApi.yml",
            ],
            testProjectName: `${this.props.githubRepo}-Tests`,
            vpc: this.props.vpc,
            env: this.props.env,
            bucketRemovalPolicy: RemovalPolicy.DESTROY,
            domain: "workshop.com",
            stackName: `${this.formatRepoNameForCloudFormation()}-CodePipeline`,
            manualApproval: false,
            shouldCreateNewSlackChannel: false,
        };

        return new PipelineStack(this.props.scope, pipeline);
    }

    private createRdsStack(): RdsStack {
        const props: RdsStackProps = {
            env: this.props.env,
            vpc: this.props.vpc,
            databaseName: "workshop",
            databaseUserName: "masteruser",
            backupRetentionDays: 14,
            storageSizeGb: 100,
            monitoringIntervalSeconds: 60,
            engineVersion: PostgresEngineVersion.VER_17_4,
            instanceType: InstanceType.of(InstanceClass.T3, InstanceSize.MICRO),
            githubRepo: this.props.githubRepo,
            envName: AppEnvironment.Production,
            stackName: `${this.formatRepoNameForCloudFormation()}-rds`,
            name: "RdsPostgres",
        };

        return new RdsStack(this.props.scope, props);
    }

    private createBastionStack(): WindowsBastionStack {
        const windowsBastionStackProps: WindowsBastionStackProps = {
            vpc: this.props.vpc,
            githubRepo: this.props.githubRepo,
            envName: AppEnvironment.Production,
            name: "bastion",
            stackName: `${this.formatRepoNameForCloudFormation()}-Bastion`,
            env: this.props.env,
        };

        return new WindowsBastionStack(
            this.props.scope,
            windowsBastionStackProps,
        );
    }
}
