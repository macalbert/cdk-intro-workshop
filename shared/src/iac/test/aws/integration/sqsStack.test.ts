import { App, Duration, Stack } from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import {
	SqsStack,
	type SqsStackProps,
} from "../../../src/aws/integration/sqsStack";
import { AppEnvironment } from "../../../src/config/env/appEnvironment";

describe("SQS Stack", () => {
	const env = {
		account: "account",
		region: "eu-west-1",
	};

	test("Should_SqsWithDeadLetter_When_StackIsCalled", () => {
		// Arrange
		const stack = new Stack(new App(), "sqsStackTest", { env: env });
		const props = createSqsStackProps(true);

		// Act
		const actual = new SqsStack(stack, props);

		// Assert
		const template = Template.fromStack(actual);
		expect(template.toJSON()).toMatchSnapshot("sqsStackTest");
	});

	test("Should_SqsWithoutDeadLetter_When_StackIsCalled", () => {
		// Arrange
		const stack = new Stack(new App(), "sqsWithoutDeadLetterStackTest", {
			env: env,
		});
		const props = createSqsStackProps(false);

		// Act
		const actual = new SqsStack(stack, props);

		// Assert
		const template = Template.fromStack(actual);
		expect(template.toJSON()).toMatchSnapshot("sqsWithoutDeadLetterStackTest");
	});

	function createSqsStackProps(deadLetter: boolean): SqsStackProps {
		return {
			env: env,
			name: "SqsStackTest",
			envName: AppEnvironment.Test,
			githubRepo: "test",
			stackName: `test-apigateway`,
			queueName: "string",
			deadLetter: deadLetter,
			maxReceiveCount: 3,
			visibilityTimeout: Duration.minutes(10),
			retentionPeriod: Duration.days(14),
		};
	}
});
