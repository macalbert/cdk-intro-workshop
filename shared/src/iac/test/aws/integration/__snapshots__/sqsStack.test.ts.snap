// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SQS Stack Should_SqsWithDeadLetter_When_StackIsCalled: sqsStackTest 1`] = `
{
  "Outputs": {
    "DeadLetterQueueArn": {
      "Value": {
        "Fn::GetAtt": [
          "DeadLetterQueue9F481546",
          "Arn",
        ],
      },
    },
    "MainQueueArn": {
      "Value": {
        "Fn::GetAtt": [
          "MainQueueD24C6076",
          "Arn",
        ],
      },
    },
    "teststringTestapiparamParameterNameOutput": {
      "Description": "The SQS url test string Test",
      "Export": {
        "Name": "test-string-Test-api-paramParameterName",
      },
      "Value": {
        "Ref": "teststringTestapiparam5B953861",
      },
    },
    "teststringTestworkerserviceparamParameterNameOutput": {
      "Description": "The SQS url test string Test",
      "Export": {
        "Name": "test-string-Test-workerservice-paramParameterName",
      },
      "Value": {
        "Ref": "teststringTestworkerserviceparamAAB3492B",
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "DeadLetterQueue9F481546": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "MessageRetentionPeriod": 1209600,
        "QueueName": "string-deadletter",
        "VisibilityTimeout": 0,
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "DeadLetterQueuePolicyB1FB890C": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "sqs:SendMessage",
                "sqs:GetQueueAttributes",
                "sqs:GetQueueUrl",
              ],
              "Effect": "Allow",
              "Principal": {
                "Service": "sqs.amazonaws.com",
              },
              "Resource": {
                "Fn::GetAtt": [
                  "DeadLetterQueue9F481546",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Queues": [
          {
            "Ref": "DeadLetterQueue9F481546",
          },
        ],
      },
      "Type": "AWS::SQS::QueuePolicy",
    },
    "MainQueueD24C6076": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "DelaySeconds": 5,
        "MessageRetentionPeriod": 1209600,
        "QueueName": "string",
        "RedrivePolicy": {
          "deadLetterTargetArn": {
            "Fn::GetAtt": [
              "DeadLetterQueue9F481546",
              "Arn",
            ],
          },
          "maxReceiveCount": 3,
        },
        "VisibilityTimeout": 600,
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "teststringTestapiparam5B953861": {
      "Properties": {
        "AllowedPattern": ".*",
        "Description": "The SQS url test string Test",
        "Name": "/M47.Test.Apps.Minimal.Api/Test/Sqsstacktest/Aws/QueueUrl",
        "Tier": "Standard",
        "Type": "String",
        "Value": {
          "Ref": "MainQueueD24C6076",
        },
      },
      "Type": "AWS::SSM::Parameter",
    },
    "teststringTestworkerserviceparamAAB3492B": {
      "Properties": {
        "AllowedPattern": ".*",
        "Description": "The SQS url test string Test",
        "Name": "/M47.Test.Apps.WorkerService/Test/Sqsstacktest/Aws/QueueUrl",
        "Tier": "Standard",
        "Type": "String",
        "Value": {
          "Ref": "MainQueueD24C6076",
        },
      },
      "Type": "AWS::SSM::Parameter",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;

exports[`SQS Stack Should_SqsWithoutDeadLetter_When_StackIsCalled: sqsWithoutDeadLetterStackTest 1`] = `
{
  "Outputs": {
    "MainQueueArn": {
      "Value": {
        "Fn::GetAtt": [
          "MainQueueD24C6076",
          "Arn",
        ],
      },
    },
    "teststringTestapiparamParameterNameOutput": {
      "Description": "The SQS url test string Test",
      "Export": {
        "Name": "test-string-Test-api-paramParameterName",
      },
      "Value": {
        "Ref": "teststringTestapiparam5B953861",
      },
    },
    "teststringTestworkerserviceparamParameterNameOutput": {
      "Description": "The SQS url test string Test",
      "Export": {
        "Name": "test-string-Test-workerservice-paramParameterName",
      },
      "Value": {
        "Ref": "teststringTestworkerserviceparamAAB3492B",
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "MainQueueD24C6076": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "DelaySeconds": 5,
        "MessageRetentionPeriod": 1209600,
        "QueueName": "string",
        "VisibilityTimeout": 600,
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "teststringTestapiparam5B953861": {
      "Properties": {
        "AllowedPattern": ".*",
        "Description": "The SQS url test string Test",
        "Name": "/M47.Test.Apps.Minimal.Api/Test/Sqsstacktest/Aws/QueueUrl",
        "Tier": "Standard",
        "Type": "String",
        "Value": {
          "Ref": "MainQueueD24C6076",
        },
      },
      "Type": "AWS::SSM::Parameter",
    },
    "teststringTestworkerserviceparamAAB3492B": {
      "Properties": {
        "AllowedPattern": ".*",
        "Description": "The SQS url test string Test",
        "Name": "/M47.Test.Apps.WorkerService/Test/Sqsstacktest/Aws/QueueUrl",
        "Tier": "Standard",
        "Type": "String",
        "Value": {
          "Ref": "MainQueueD24C6076",
        },
      },
      "Type": "AWS::SSM::Parameter",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
